components/brave_news/browser/suggestions_controller.cc

```diff
@@ -155,25 +156,18 @@ void SuggestionsController::GetSuggestedPublisherIds(
                     [](SuggestionsController* controller,
                        GetSuggestedPublisherIdsCallback callback,
                        Publishers publishers) {
-                      history::QueryOptions options;
-                      options.max_count = 2000;
-                      options.SetRecentDayRange(14);
-                      controller->history_service_->QueryHistory(
-                          std::u16string(), options,
-                          base::BindOnce(
-                              [](SuggestionsController* controller,
-                                 Publishers publishers,
-                                 GetSuggestedPublisherIdsCallback callback,
-                                 history::QueryResults results) {
-                                auto result =
-                                    controller
-                                        ->GetSuggestedPublisherIdsWithHistory(
-                                            publishers, results);
-                                std::move(callback).Run(std::move(result));
-                              },
-                              base::Unretained(controller),
-                              std::move(publishers), std::move(callback)),
-                          &controller->task_tracker_);
+                      controller->history_querier_->Run(base::BindOnce(
+                          [](SuggestionsController* controller,
```

Comment:
I'm not 100% sure, but we may need to use weak ptr for `SuggestionController`.

My theroy is that the new engine will be destroyed first and then the task runner will be, so suggestion controller might be a dangling pointer. Although we're checking the task is canceled, but I guess that could still be racey, when the callback to main thread is already posted.

Reply:
I think you might be right! Nice catch